swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1292,7 +1403,111 @@ string ppReprType(DataFlowType t) { none() }
  * a node of type `t1` to a node of type `t2`.
  */
 pragma[inline]
-predicate compatibleTypes(DataFlowType t1, DataFlowType t2) { any() }
+predicate compatibleTypes(DataFlowType t1, DataFlowType t2) {
+  subType(t1) = subType(t2)
+  or
+  exists(BoundGenericType bound1, BoundGenericType bound2 |
+    stripType(t1.asType()).(BoundGenericType) = bound1 and
+    stripType(t2.asType()).(BoundGenericType) = bound2 and
+    bound1.getDeclaration() = bound2.getDeclaration() and
+    forall(int index | index in [0 .. bound1.getNumberOfArgTypes() - 1] |
+      isAnyOrUnboundType(bound1.getArgType(index).getCanonicalType()) or
+      isAnyOrUnboundType(bound2.getArgType(index).getCanonicalType())
+    )
+  )
+  or
+  exists(TupleType tuple1, TupleType tuple2 |
+    stripType(t1.asType()) = tuple1 and
+    stripType(t2.asType()) = tuple2 and
+    compatibleTuples(tuple1, tuple2)
+  )
+  or
+  t1 instanceof TopFunctionType and
+  t2 instanceof DataFlowFunctionType
+  or
+  t1 instanceof DataFlowFunctionType and
+  t2 instanceof TopFunctionType
+  or
+  t1 instanceof TopFunctionType and
+  t2 instanceof TopFunctionType
+  or
+  t1.asCallable() = t2.asCallable()
+  or
+  isAnyOrUnboundType(stripType(t1.asType()))
+  or
+  isAnyOrUnboundType(stripType(t2.asType()))
+}
+
+pragma[nomagic]
+DataFlowType subType(DataFlowType t) {
+  result.asType().getABaseType*().getCanonicalType() = stripType(t.asType())
+}
+
+predicate isAnyOrUnboundType(Type t) {
+  t instanceof AnyType
+  or
+  t instanceof GenericTypeParamType
+  or
+  t instanceof ArchetypeType
+  or
+  t instanceof DependentMemberType
+  or
+  exists(TypeDecl td, TypeDecl parent |
+    t.(AnyGenericType).getDeclaration() = td and
+    parent = getEnclosingDeclNonExtension+(td) and
+    (
+      parent.(GenericTypeDecl).getNumberOfGenericTypeParams() != 0 or
+      parent.(ProtocolDecl).getNumberOfGenericTypeParams() != 0
+    )
+  )
+}
+
+pragma[inline]
+predicate compatibleTuples(TupleType tuple1, TupleType tuple2) {
+  compatibleTuplesPartial(tuple1, tuple2, tuple1.getNumberOfTypes() - 1)
+}
+
+predicate compatibleTuplesPartial(TupleType tuple1, TupleType tuple2, int index) {
+  (
+    index = 0 and
+    tuple1.getNumberOfTypes() = tuple2.getNumberOfTypes()
+    or
+    compatibleTuplesPartial(tuple1, tuple2, index - 1)
+  ) and
+  (
+    isAnyOrUnboundType(tuple1.getType(index).getCanonicalType()) or
+    isAnyOrUnboundType(tuple2.getType(index).getCanonicalType())
+  )
+}
```

Comment:
It almost looks like what we want to compute here is a GVN of a tuple type 🤔 (which is actually what C# does: https://github.com/github/codeql/blob/main/csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll#L49)